### DISCUSSION TITLE
How to remove small background objects

### DISCUSSION THREAD

**Discussion by luisfl** [ImageMagick/ImageMagick#3702] [Digital Image Processing] [answered]
I have images with letters that I want to remove. From what I've seen IM do, I am sure this can be done with **convert**, there should even be well known tricks for this, but **convert** is too complex for this newbie. 

So, suppose this is the image:

![CarinaNebula_NGC3372_52stack_GIMP_LessStars_JN-210416-B_59577a1](https://user-images.githubusercontent.com/15525268/118821687-dd70f880-b88d-11eb-9eea-a9b1e85e9591.jpg)

What I want is an image with the "Eta Carina" text removed.

What I did so far was to copy a rectangle from other part of the image and merge it to the original one. I also did a blur at the boundaries of the copied rectangle to make the seams less obvious. Still, depending on the image, it is still obvious (not really in this image that is easy).

So I am trying now to "remove the letters and fill the holes after blurring them". So far I got an image with transparent holes, like this:

`convert -compress None carinae.jpg carinaerectangle.jpg -fuzz 30% -compose ChangeMask -composite -channel A -negate output.png`

Now maybe I should enlarge the holes, blur them, and merge with the rectangle? Or some other way?

This should be more or less standard...?

Thanks!

L.

**Comment by snibgo** [comment, marked as answer]
After making those pixels transparent, you could use any of the methods I show on [Filling holes in priority order](http://im.snibgo.com/fillholespri.htm) or [Filling holes](http://im.snibgo.com/fillholes.htm).

However, those methods assume the image contains detail that should be replicated in the filled area. A better solution might be to simply flatten against black or a suitable near-black.

**Reply by luisfl** [reply]
Yes, I saw those modules. Pretty impressive! I should learn how to install them (it seems there are no RPM packages of them for Fedora...)

Regarding your "better solution", that would indeed work for this particular image, but in general I have gradient in the background, or a focus blurred background, in which case a solid color fill looks terrible. In general, I do have other parts of the image where I can clone stuff.

**Reply by luisfl** [reply]
I successfully compiled your modules. I run it successfully as in your example:

        `convert  cu_fh_src1.png  -process 'fillholes wr 2'  cu_fh_src1_fh.png`

However, when I tried on this PNG image (the white rectangle is actually transparent)

![mia](https://user-images.githubusercontent.com/15525268/118860039-4b7ae700-b8b1-11eb-8a9d-e1168001ba97.png)


it took forever, apparently without ending. Not sure what I did wrong.

**Comment by snibgo** [comment]
See the web pages for performance hints, especially the "lsr" option.

For example, three iterations of the following takes 1.5 minutes:
```
magick x2.png -process 'fillholespri wr 2 lsr 5c copy window v' x2.png
```
![x2](https://user-images.githubusercontent.com/33812368/118865764-3619b300-b8d9-11eb-82f6-9c74b7a0c296.png)

**Reply by luisfl** [reply]
Ok, now I got a similar result to you, but in 30 seconds. Thanks!

**Comment by fmw42** [comment]
Here is one way using morphology top-hat to get a mask of the letters. Then use morphology open to process the image and blend that with the original using the mask of the letter that are dilated.

Input:

![eta_carinae](https://user-images.githubusercontent.com/13458218/118865886-4b212280-b896-11eb-955c-8123dcc7b248.jpg)

```
convert eta_carinae.jpg \
\( -clone 0 -morphology open diamond:4 \) \
\( -clone 0 -colorspace gray \
-morphology tophat:5 diamond:2 \
-auto-level \
-negate -threshold 13% -negate +write eta_carinae_text.png \
-morphology dilate diamond:2 \) \
-compose over -composite \
eta_carinae_cleaned.png
```

Letters:

![eta_carinae_text](https://user-images.githubusercontent.com/13458218/118865942-583e1180-b896-11eb-856a-a60e501843ad.png)

Result:

![eta_carinae_cleaned](https://user-images.githubusercontent.com/13458218/118865974-5ecc8900-b896-11eb-97c1-0e48452611da.png)

**Reply by luisfl** [reply]
Thanks for this.
I see that you left a black hole in place. I need more like a clone thing (without adjusting parameters), and with blurred nonblack background. Probably my example was a bad one... See the other PNG image with a hole.

**Comment by snibgo** [comment]
I wrote:
> For example, three iterations of the following ...

I forgot that both "fillholes" and "fillholespri" both now have an "auto-repeat" option, so there is no need to run the command multiple times.

**Reply by luisfl** [reply]
For my installation, fillholespri works great with "auto-repeat", but fillholes ignores it.

**Comment by luisfl** [comment]
Thanks @snibgo . Your algorithms are impressive indeed. However, for my uses, they just take too long, and leave certain artifacts that are quite visible for out of focus background.

Let me try again. Suppose that I have this image, that I want to remove the lower left white text:

![sai-verde_Ubatumirim-SP-210108-B_54615a](https://user-images.githubusercontent.com/15525268/118897717-3f0e8280-b8e1-11eb-9b0d-c7ce7c6a8d10.jpg)

So I copied a rectangle from above over it:

![sai-verde_Ubatumirim-SP-210108-B_54615a_v](https://user-images.githubusercontent.com/15525268/118897761-53527f80-b8e1-11eb-9689-b90cfeb9707f.jpg)

I can do a bit better blurring the borders of the rectangle, but still quite obvious:

![sai-verde_Ubatumirim-SP-210108-B_54615a_V](https://user-images.githubusercontent.com/15525268/118897809-69f8d680-b8e1-11eb-9834-2f2150f45ace.jpg)

So I am now trying this to generate the output rectangle ($outfilemerged) that I then paste over the original image (the -crop ${cropsize}+${originx}+${originy} \"$outfile\" refers to the rectangle above the letters, and $outfile is the same sized rectangle containing the letters):

     `convert -compress None \"$outfile\" \\( \"$file\" -crop ${cropsize}+${originx}+${originy} \"$outfile\" -fuzz 30% -compose ChangeMask -composite -channel A -negate -morphology Erode Octagon -morphology Distance Euclidean:4,5%\\) -compose atop -composite \"$outfilemerged\"`

This is what I got:

![sai-verde_Ubatumirim-SP-210108-B_54615a_b](https://user-images.githubusercontent.com/15525268/118898018-fe633900-b8e1-11eb-82da-ce2cc93b9b2f.jpg)

Still not good....

For comparison, this is the result after cropping a rectangle containing the letters and running

       `convert $file  -process 'fillholes a wr 2 lsr 5c copy window v' output.jpg`

![output2](https://user-images.githubusercontent.com/15525268/118899388-02dd2100-b8e5-11eb-9f1f-27697f376d0f.jpg)

Indeed looks better than all my attempts. It took 5 seconds only.

L.

**Comment by snibgo** [comment]
fillholes and fillholespri create discontinuity at the centre of the holes, so they work best with small holes. (Other methods don't have the same discontinuity.)

So we make just the letters transparent, and fill them. Windows BAT script:

```
%IM7DEV%magick ^
  saiverde.jpg ^
  ( +clone ^
    -crop 381x95+65+879 ^
    -alpha opaque ^
    -fuzz 5%% ^
    -fill Black -opaque White ^
    -fill White +opaque Black ^
    -morphology erode disk:1 ^
    -geometry %%X%%Y ^
  ) ^
  -alpha off -compose CopyOpacity -composite ^
  -process 'fillholes a wr 2 lsr 5c copy window v' ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/118903244-42206780-b90f-11eb-8604-1a5e32c442e7.png)

**Reply by luisfl** [reply]
Spectacular!!! Could you please explain me what your command is actually doing in each step (after the alpha opaque)? "-fill White +opaque Black"? Thanks!!

EDIT: if the input file is JPG I get an error (but it works for PNG):

      `convert: no decode delegate for this image format `JPG' @ error/constitute.c/ReadImage/572.`

and I also get an error at the -geometry line:

      `convert: invalid argument for option '-geometry': %%X%%Y @ error/convert.c/ConvertImageCommand/1746.`

Removing the -geometry line, I get back the original file.

**Comment by snibgo** [comment]
What version of IM do you use? If v7, then use "magick" instead of "convert".

If you use v6, my command won't work (because of "-geometry"), so we'll think of something else

What shell do you use? I showed a Windows BAT command. If you use bash, then change the line-ends from caret ^ to backslash \, and change double %% to single %, and change ( and ) to \( and \).
```
%IM7DEV%magick ^ <== run "magick"
  saiverde.jpg ^ <== read your bird image
  ( +clone ^ <== start a new list, cloning the bird image
    -crop 381x95+65+879 ^ <== crop an rectangle that encloses the text
    -alpha opaque ^ <== enable alpha channel, making all pixels opaque
    -fuzz 5%% ^  <== set fuzz amount, because the "white" text isn't exactly white
    -fill Black -opaque White ^ <== change white pixels to black
    -fill White +opaque Black ^ <== change non-black pixels to white. The letters are now black on a white background.
    -morphology erode disk:1 ^ <== erode the white, ie dilate the black leters, making them slightly larger
    -geometry %%X%%Y ^ <== "-crop" created page offsets. Copy these to "-geometry", for "-composite".
  ) ^ <== close the image list, copying the result image to the outer list.

<== Now we have two images: the bird, and a small rectangle with enlarged black text on white background.

  -alpha off -compose CopyOpacity -composite ^ <== Use the black enlarged text as transparency.

<== Phew! Now we have transparent text.

  -process 'fillholes a wr 2 lsr 5c copy window v' ^ <== fill the holes
  x.png  <== write the output
```

Why do we crop? Because other parts of the image are white, and we don't want those to become transparent.

**Reply by luisfl** [reply]
My dev environment is V7, I failed to install your plugins in V6 (6.9.12). I use BASH in Linux. I fixed the caret, \( \), and such already. The problem was the double %%, now it is working.

Yet, for some reason it is not accepting JPGs as input, and even when saving as x.jpg, the saved file is indeed PNG... Something should be wrong with my DEV IM. `magick -list format` doesn't show any JPEG/JPG. The `.configure` gave "no" at all JPEG stuff. Maybe I have some devel JPEG library missing?

Thank you VERY much for your explanation!! Very instructive and clever.

Your algorithm works for white letters only. It would be great if the letters can be identified by contrast or something. That was my idea when I tried to compare with a rectangle above the letters. But my idea is strongly dependent on the image also, so it does not work well.

Thanks again!

**Comment by snibgo** [comment]
> Maybe I have some devel JPEG library missing?

Yes, I expect so. Did you build IM yourself? If so, you should have a config.log, that may say why JPEG isn't included.

> Your algorithm works for white letters only. It would be great if the letters can be identified by contrast or something.

Algorithms exist for finding text, for example car number plates ("licence plates") in photos. This case is simpler, and you could search for groups of adjacent identical pixels. But your input has (lossy) JPEG compression, which changes pixels values so they are no longer identical.

**Reply by luisfl** [reply]
Yes, I had libjpeg missing. Recompiled and working now.
I build IM since your plugins are not included.

I will think about those algorithms you suggest.

Thanks!
